blog/_posts/2022-03-15-pulumi.md

```diff
@@ -0,0 +1,275 @@
+---
+slug: "pulumi"
+date: 2022-03-16
+author: nick
+layout: blog
+title: "Local Kubernetes Development With Pulumi"
+subtitle: "How to develop in containers without messing around with YAML"
+description: "How to develop in containers without messing around with YAML."
+image: "/assets/images/pulumi/platypus.jpg"
+image_caption: "This platypus isn't as purple as the Pulumi mascot but it's still cute. Courtesy of <a href='https://oceanservice.noaa.gov/facts/platypus.html'>NOAA</a>."
+tags:
+  - pulumi
+  - yaml
+  - kubernetes
+---
+
+Kubernetes is almost 8 years old now. I like Kubernetes a lot! But one of the
+big surprises is that users are still messing around with YAML files. This has
+surprised a few other people in the community too:
+
+<blockquote class="twitter-tweet"><p lang="en" dir="ltr">As <a href="https://twitter.com/bryanl?ref_src=twsrc%5Etfw">@bryanl</a> says: YAML is for computers. When we started with YAML we never intended it to be the user facing solution. We saw it as &quot;assembly code&quot;. I&#39;m horrified that we are still interacting with it directly. That is a failure.</p>&mdash; Joe Beda (@jbeda) <a href="https://twitter.com/jbeda/status/994566252503810048?ref_src=twsrc%5Etfw">May 10, 2018</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+
+And Joe's tweet above is from 4 years ago!
+
+In the past, we've written about the ways you can use
+[Helm](https://blog.tilt.dev/2022/01/26/helm-improvements.html) or
+[Kustomize](https://blog.tilt.dev/2020/02/04/are-you-my-kustomize.html). Both
+these tools are great for organizing YAML into packages and adapting them to
+multiple environments. Helm is usually the first thing that I personally reach
+for when my YAML is getting unwieldy. But they are the "buy a USB hub"
+solution to infrastructure.
+
+I've been playing around with Pulumi recently as an alternative.
+
+Pulumi generates real programming language APIs for infrastructure, so you can
+set up your infrastructure with code rather than with YAML. Then you can
+organize the common bits in the way that works best for your services.
+
+It's still declarative! It's just defined in code. And this lets Pulumi provide more of the building blocks of a nice, pleasant-to-use deploy system:
+
```

Comment:
Possibly worth mentioning here that Tilt follows a similar idea, which is why tiltfiles are written in starlark rather than config files.